### PR TITLE
DPE-2959 async replication

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Need to have async replication

## Solution

Use async replication library (owned by vm charm).
Implementation as described by this [workflow](https://miro.com/app/board/uXjVN9INa2Q=/)
How-to draft in the [discourse page](https://discourse.charmhub.io/t/charmed-mysql-k8s-async-replication/12904/1)

Also implements [DPE-3761](https://warthogs.atlassian.net/browse/DPE-3761)

Integration test included

### TODO:

* ~minimal unit test fixes~
* ~discourse doc page~
* ~full usage of data-platform secrets~
* test endpoint reachability
* ~fence write traffic for old primary~

_Unit tests tests on a fup PR on VM only(lib owner)_

[DPE-3761]: https://warthogs.atlassian.net/browse/DPE-3761?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by taurus-forever** [comment]
As discussed in private: it is not possible to re-join the cluster after `recreate-cluster`. The juju stuck on:
```
Unit    Workload  Agent  Address       Ports  Message
db2/0*  blocked   idle   10.1.204.228         User data found, aborting async replication setup
db2/1   waiting   idle   10.1.204.196         waiting replica cluster be configured
db2/2   waiting   idle   10.1.204.207         waiting replica cluster be configured
```

STR:
1. https://discourse.charmhub.io/t/charmed-mysql-k8s-async-replication/12904
2. After `juju run -m az2 db2/leader recreate-cluster` it is not possible to re-join db2 to az1.db1 because az2 has the previos data which is impossible to wipe.

Proposal:
1) the blocked cluster after  `juju remove-relation -m az2 async-primary db2` can be re-joined back using `juju relate -m az2 async-primary db2` (the local az2 data will be wiped)
2) if the cluster re-create, it cannot re-join, it must be destroyed and re-deployed. Safety first.

**Comment by paulomach** [comment]
@taurus-forever 

> 
>     2. if the cluster re-create, it cannot re-join, it must be destroyed and re-deployed. Safety first.

Should we make this parametrized for the action?

**Comment by taurus-forever** [comment]
> @taurus-forever
> 
> > ```
> > 2. if the cluster re-create, it cannot re-join, it must be destroyed and re-deployed. Safety first.
> > ```
> 
> Should we make this parametrized for the action?

IMHO, not now. It is a dangerous action. If you want to re-attach the re-created cluster: 
`juju remove-application mysql && juju remove-storage ... && juju deploy mysql` 
(c) Safety first.

P.S. we can implement such action later after the proper spec review by PMs.

**Comment by taurus-forever** [comment]
For the history, revision 128 from edge/arepl fails:

```
unit-db2-0: 14:13:57 DEBUG unit.db2/0.juju-log async-replica:3: Syncing credentials from primary cluster                   
unit-db2-0: 14:13:57 ERROR unit.db2/0.juju-log async-replica:3: Uncaught exception while in charm code:                    
Traceback (most recent call last):                                                                                         
  File "/var/lib/juju/agents/unit-db2-0/charm/venv/ops/model.py", line 3019, in _run                                       
    result = subprocess.run(args, **kwargs)  # type: ignore                                                                
  File "/usr/lib/python3.10/subprocess.py", line 526, in run                                                               
    raise CalledProcessError(retcode, process.args,                                                                        
subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-db2-0/secret-get', 'secret:cnpfe3nmp25c77tiud60', '--format=json')' returned non-zero exit status 1.
```

**Comment by paulomach** [comment]
> For the history, revision 128 from edge/arepl fails:
> 
> ```
> unit-db2-0: 14:13:57 DEBUG unit.db2/0.juju-log async-replica:3: Syncing credentials from primary cluster                   
> unit-db2-0: 14:13:57 ERROR unit.db2/0.juju-log async-replica:3: Uncaught exception while in charm code:                    
> Traceback (most recent call last):                                                                                         
>   File "/var/lib/juju/agents/unit-db2-0/charm/venv/ops/model.py", line 3019, in _run                                       
>     result = subprocess.run(args, **kwargs)  # type: ignore                                                                
>   File "/usr/lib/python3.10/subprocess.py", line 526, in run                                                               
>     raise CalledProcessError(retcode, process.args,                                                                        
> subprocess.CalledProcessError: Command '('/var/lib/juju/tools/unit-db2-0/secret-get', 'secret:cnpfe3nmp25c77tiud60', '--format=json')' returned non-zero exit status 1.
> ```

Updated revision r129 with fixes

File: lib/charms/mysql/v0/async_replication.py [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+    def _on_primary_created(self, event):
+        """Validate relations and share credentials with replica cluster."""
+        if not self._charm.unit_initialized:
+            logger.debug("Unit not initialized, deferring event")
+            event.defer()
+            return
+
+        if self._charm._mysql.is_cluster_replica():
+            logger.error(
+                f"This a replica cluster, cannot be related as {PRIMARY_RELATION}. Remove relation."
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/async_replication.py:361]: This `is` a replica cluster

File: lib/charms/mysql/v0/async_replication.py [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+            logger.debug("Initializing replica cluster")
+            self._charm.unit.status = MaintenanceStatus("Adding replica cluster")
+            remote_data = self.get_remote_relation_data(event.relation) or {}
+
+            cluster = remote_data["cluster-name"]
+            endpoint = remote_data["endpoint"]
+            unit_label = remote_data["node-label"]
+
+            if cluster == self.cluster_name:
+                import uuid
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/async_replication.py:404]: avoid inline imports

File: lib/charms/mysql/v0/async_replication.py [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+            )
+        else:
+            # non-leader/secondaries
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_created,
+                self._on_replica_secondary_created,
+            )
+            self.framework.observe(
+                self._charm.on[REPLICA_RELATION].relation_changed,
+                self._on_replica_secondary_changed,
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/async_replication.py:460]: this is a bit confusing: should secondary be renamed to `non_leader` here if it is in juju sense?

File: lib/charms/mysql/v0/async_replication.py [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+                    "User data found, check instruction in the log"
+                )
+                self._charm.unit.status = BlockedStatus(
+                    "User data found, aborting async replication setup"
+                )
+                self.relation_data["user-data-found"] = "true"
+                return
+
+        self._charm.app.status = MaintenanceStatus("Setting up async replication")
+        self._charm.unit.status = WaitingStatus("awaiting sync data from primary cluster")
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/async_replication.py:594]: awaiting -> Awaiting

File: lib/charms/mysql/v0/async_replication.py [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+                self._charm.app_peer_data["cluster-set-domain-name"] = cluster_set_domain_name
+
+            # set the number of units added to the cluster for a single unit replica cluster
+            # needed here since it will skip the `RECOVERING` state
+            if self._charm.app.planned_units() == 1:
+                self._charm.app_peer_data["units-added-to-cluster"] = "1"
+
+            self._charm._on_update_status(None)
+        elif state == States.RECOVERING:
+            # recoveryng cluster (copying data and/or joining units)
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/async_replication.py:690]: recoveryng -> recovering

File: src/charm.py [hunk 6]
```diff
@@ -276,55 +283,71 @@ def join_unit_to_cluster(self) -> None:
...
 
-            if self._mysql.get_cluster_node_count(from_instance=cluster_primary) == GR_MAX_MEMBERS:
-                self.unit.status = WaitingStatus(
-                    f"Cluster reached max size of {GR_MAX_MEMBERS} units. Standby."
-                )
-                logger.warning(
-                    f"Cluster reached max size of {GR_MAX_MEMBERS} units. This unit will stay as standby."
+                # add random delay to mitigate collisions when multiple units are joining
+                # due the difference between the time we test for locks and acquire them
+                sleep(random.uniform(0, 1.5))
```
Comment (shayancanonical) [inline src/charm.py:319]: why do we need this sleep? isnt getting a lock atomic in MySQL the way we acquire them?

**Review by shayancanonical — COMMENTED** [review]
Preliminary pass - intend to do a deeper dive in the coming days

File: lib/charms/mysql/v0/async_replication.py [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+        )
+
+    @cached_property
+    def role(self) -> ClusterSetInstanceState:
+        """Current cluster set role of the unit, after the relation is established."""
+        is_replica = self._charm._mysql.is_cluster_replica()
+
+        if is_replica:
+            cluster_role = "replica"
+        elif is_replica is False:
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:95]: comparison `False` with is is suitable, given that `False` is one of builtin singletons

File: lib/charms/mysql/v0/async_replication.py [hunk 8]
```diff
@@ -0,0 +1,722 @@
...
+                logger.info("Fencing writes to the cluster")
+                self._charm._mysql.fence_writes()
+                event.set_results({"message": "Writes to the cluster are now fenced"})
+            elif (
+                event.handle.kind == "unfence_writes_action"
+                and self._charm._mysql.is_cluster_writes_fenced()
+            ):
+                logger.info("Unfencing writes to the cluster")
+                self._charm._mysql.unfence_writes()
+                event.set_results({"message": "Writes to the cluster are now resumed"})
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:175]: Fencing state is shown on the status message, hence not bothering doing that here.

File: lib/charms/mysql/v0/async_replication.py (hunk 2 above)
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py:404]: ops, forgot this one

File: src/charm.py (hunk 6 above)
Comment (paulomach) [reply src/charm.py:319]: It's a hack to have less collisions on testing for the lock. I've noticed that, when triggered together, tries to acquire the lock takes longer to return then when the lock is committed with the random delay, rendering the process a bit snappier

File: lib/charms/mysql/v0/async_replication.py (hunk 3 above)
Comment (paulomach) [reply lib/charms/mysql/v0/async_replication.py]: Agreed. Changed

**Review by taurus-forever — APPROVED** [review]
Let's move forward here, to test further in edge and waiting for Oracle fix for MySQL.

**Pull request merged**